### PR TITLE
Support  `_exportId` for bulk exports.

### LINKED ISSUE #4774: mark binaries generated by bulk export with useful metadata.
Currently, upon reading a binary resource, it is impossible to tell what generated it. This feature requests: 

1. That the bulk export provider adds a new optional parameter, `_exportId` to its initiation methods, which will be used to mark resources generated by that bulk export.
2. that the WriteBinaryFiles step annotates its generated binaries with: 
   - the provided `_exportId` if available. 
   - the job ID of the job instance. 
   - the resource type of the data contained within that binary. 
3. Since Binaries do not support extensions, store this data in the `binary.meta.extensions` field.

### PR DESCRIPTION
**Pull request by tadgh**
Closes #4774 

- New Pointcut for prefixing binaries which permits implementers to prefix a blob ID.
- New operation param for Bulk Export to allow it to pass in an `_exportId`
- Tag all bulk export-generated resources with the job id, export it, and resource type of the creating job.

### CODE REVIEW

File: hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/export/WriteBinaryStep.java [hunk 1]
```diff
@@ -122,6 +130,39 @@ public RunOutcome run(@Nonnull StepExecutionDetails<BulkExportJobParameters, Exp
...
 
+	/**
+	 * Adds 3 extensions to the `binary.meta` element.
+	 *
+	 * 1. the _exportId provided at request time
+	 * 2. the job_id of the job instance.
+	 * 3. the resource type of the resources contained in the binary
+	 */
+	private void addMetadataExtensionsToBinary(@Nonnull  StepExecutionDetails<BulkExportJobParameters, ExpandedResourcesList> theStepExecutionDetails, ExpandedResourcesList expandedResources, IBaseBinary binary) {
+		if (binary.getMeta() instanceof IBaseHasExtensions) {
```
Comment (jamesagnew) [inline hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/export/WriteBinaryStep.java:141]:
```suggestion
		// Note that this applies only to hl7.org structures, so these extensions will not be added
		// to DSTU2 structures
		if (binary.getMeta() instanceof IBaseHasExtensions) {
```

File: hapi-fhir-storage-batch2-jobs/src/test/java/ca/uhn/fhir/batch2/jobs/export/BulkExportJobParametersValidatorTest.java [hunk 2]
```diff
@@ -55,6 +55,35 @@ public void validate_validParametersForSystem_returnsEmptyList() {
...
+		parameters.setExportIdentifier("exportId&&&");
+		// when
+		when(myDaoRegistry.isResourceTypeSupported(anyString()))
+			.thenReturn(true);
+		List<String> errors = myValidator.validate(parameters);
+
+		// verify
+		assertNotNull(errors);
+		assertEquals(1, errors.size());
+		assertEquals(errors.get(0), "Export ID must be alphanumeric and can only contain the following special characters: *  ' ( ) _ - . / ");
```
Comment (jamesagnew) [inline hapi-fhir-storage-batch2-jobs/src/test/java/ca/uhn/fhir/batch2/jobs/export/BulkExportJobParametersValidatorTest.java:71]: Shouldn't this say "...can not contain..."?

File: hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/export/BulkExportJobParametersValidator.java [hunk 3]
```diff
@@ -69,6 +76,13 @@ public List<String> validate(@Nonnull BulkExportJobParameters theParameters) {
 		if (!Constants.CT_FHIR_NDJSON.equalsIgnoreCase(theParameters.getOutputFormat())) {
 			errorMsgs.add("The only allowed format for Bulk Export is currently " + Constants.CT_FHIR_NDJSON);
 		}
+		// validate the exportId
+		if (!StringUtils.isBlank(theParameters.getExportIdentifier())) {
+
+			if (!myBinaryStorageSvc.isValidBlobId(theParameters.getExportIdentifier())) {
```
Comment (jamesagnew) [inline hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/export/BulkExportJobParametersValidator.java:82]: Ooh this is a great idea calling this here!

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/binary/interceptor/BinaryStorageInterceptor.java [hunk 4]
```diff
@@ -206,6 +219,12 @@ private void extractLargeBinaries(TransactionDetails theTransactionDetails, IBas
 					} else {
 						assert thePointcut == Pointcut.STORAGE_PRESTORAGE_RESOURCE_CREATED : thePointcut.name();
 						newBlobId = myBinaryStorageSvc.newBlobId();
+
+						String prefix = invokeAssignBlobPrefix(theRequestDetails, theResource);
+						if (isNotBlank(prefix)) {
+							newBlobId = prefix + newBlobId;
+						}
+						myBinaryStorageSvc.isValidBlobId(newBlobId);
```
Comment (jamesagnew) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/binary/interceptor/BinaryStorageInterceptor.java:227]: Is this deliberate? We're ignoring the return value

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/binstore/FilesystemBinaryStorageSvcImpl.java [hunk 5]
```diff
@@ -75,6 +77,16 @@ private void createBasePathDirectory() {
 		mkdir(myBasePath);
 	}
 
+	/**
+	 * This implementation prevents: \ / | .
+	 */
+	@Override
+	public boolean isValidBlobId(String theNewBlobId) {
+		Pattern pattern = Pattern.compile("[\\/\\\\\\|\\.]");
```
Comment (jamesagnew) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/binstore/FilesystemBinaryStorageSvcImpl.java:85]: This method would probably read easier as:
```java
return StringUtils.containsAny(theNewBlobId, '\', '/', '|', '.');
```

**Review by jamesagnew — APPROVED** [review]

File: hapi-fhir-storage-batch2-jobs/src/test/java/ca/uhn/fhir/batch2/jobs/export/BulkExportJobParametersValidatorTest.java (hunk 2 above)
Comment (tadgh) [reply hapi-fhir-storage-batch2-jobs/src/test/java/ca/uhn/fhir/batch2/jobs/export/BulkExportJobParametersValidatorTest.java:71]: Nope, those are all actually legal characters for S3, pulled from their website. That said, this whole section will be moved to CDR as an S3-specific limitation

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/binary/interceptor/BinaryStorageInterceptor.java (hunk 4 above)
Comment (tadgh) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/binary/interceptor/BinaryStorageInterceptor.java:227]: yeah im real dumb

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/binstore/FilesystemBinaryStorageSvcImpl.java (hunk 5 above)
Comment (tadgh) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/binstore/FilesystemBinaryStorageSvcImpl.java:85]: Fair, modifying now.

**Pull request merged**